File: backend/src/models/notification.model.ts
import { DataTypes, Model } from 'sequelize';
import { sequelize } from '../config/database';
import { User } from './user.model';
import { Agency } from './agency.model';

export class Notification extends Model {
  declare public id: number;
  declare public notificationCode: string;
  declare public userId: number;
  declare public agencyId: number | null;
  declare public title: string;
  declare public message: string;
  declare public type: 'collaboration' | 'deal' | 'commission' | 'subscription' | 'system';
  declare public actionRoute: string | null;
  declare public metadata: Record<string, any> | null;
  declare public isRead: boolean;
  declare public readAt: Date | null;
  declare public channels: string;
  declare public fcmMessageId: string | null;
  declare public readonly createdAt: Date;
  declare public readonly updatedAt: Date;
}

Notification.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    notificationCode: {
      type: DataTypes.STRING(32),
      allowNull: false,
    },
    userId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: User,
        key: 'id',
      },
    },
    agencyId: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: Agency,
        key: 'id',
      },
    },
    title: {
      type: DataTypes.STRING(160),
      allowNull: false,
    },
    message: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    type: {
      type: DataTypes.ENUM('collaboration', 'deal', 'commission', 'subscription', 'system'),
      defaultValue: 'system',
    },
    actionRoute: {
      type: DataTypes.STRING(255),
      allowNull: true,
    },
    metadata: {
      type: DataTypes.JSON,
      allowNull: true,
    },
    isRead: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    readAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    channels: {
      type: DataTypes.STRING(64),
      defaultValue: 'in_app',
    },
    fcmMessageId: {
      type: DataTypes.STRING(128),
      allowNull: true,
    },
  },
  {
    tableName: 'notifications',
    sequelize,
  }
);

Notification.belongsTo(User, { foreignKey: 'userId', as: 'recipient' });
Notification.belongsTo(Agency, { foreignKey: 'agencyId', as: 'agency' });
User.hasMany(Notification, { foreignKey: 'userId', as: 'notifications' });

File: lib/features/notifications/domain/models/notification_model.dart
class AppNotification {
  final int id;
  final String notificationCode;
  final int userId;
  final int? agencyId;
  final String title;
  final String message;
  final String type; // 'collaboration', 'deal', 'commission', 'subscription', 'system'
  final String? actionRoute;
  final Map<String, dynamic>? metadata;
  final bool isRead;
  final DateTime? readAt;
  final String channels;
  final DateTime createdAt;

  const AppNotification({
    required this.id,
    required this.notificationCode,
    required this.userId,
    this.agencyId,
    required this.title,
    required this.message,
    required this.type,
    this.actionRoute,
    this.metadata,
    required this.isRead,
    this.readAt,
    required this.channels,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    DateTime parsedDate;
    try {
      parsedDate = DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String());
    } catch (_) {
      parsedDate = DateTime.now();
    }

    DateTime? parsedReadAt;
    if (json['readAt'] != null) {
      try {
        parsedReadAt = DateTime.parse(json['readAt']);
      } catch (_) {}
    }

    return AppNotification(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id'].toString()) ?? 0,
      notificationCode: json['notificationCode'] ?? '',
      userId: json['userId'] is int ? json['userId'] : int.tryParse(json['userId'].toString()) ?? 0,
      agencyId: json['agencyId'] != null ? (json['agencyId'] is int ? json['agencyId'] : int.tryParse(json['agencyId'].toString())) : null,
      title: json['title'] ?? '',
      message: json['message'] ?? '',
      type: json['type'] ?? 'system',
      actionRoute: json['actionRoute'],
      metadata: json['metadata'] is Map<String, dynamic> ? json['metadata'] : null,
      isRead: json['isRead'] == true,
      readAt: parsedReadAt,
      channels: json['channels'] ?? 'in_app',
      createdAt: parsedDate,
    );
  }

  AppNotification copyWith({
    bool? isRead,
    DateTime? readAt,
  }) {
    return AppNotification(
      id: id,
      notificationCode: notificationCode,
      userId: userId,
      agencyId: agencyId,
      title: title,
      message: message,
      type: type,
      actionRoute: actionRoute,
      metadata: metadata,
      isRead: isRead ?? this.isRead,
      readAt: readAt ?? this.readAt,
      channels: channels,
      createdAt: createdAt,
    );
  }
}
